packages/vite/src/node/optimizer/registerMissing.ts

```diff
@@ -100,49 +148,70 @@ export function createMissingImporterRegisterFn(
+      // for large applications. Comparing their fileHash we can find out if it is safe to
+      // keep the current browser state.
+      const needsReload =
+        metadata.hash !== newData.hash ||
+        Object.keys(metadata.optimized).some((dep) => {
+          return (
+            metadata.optimized[dep]?.fileHash !==
+            newData.optimized[dep]?.fileHash
+          )
+        })
 
       const commitProcessing = () => {
         processingResult.commit()
 
-        newData = optimizeDeps.metadata
+        // While optimizeDeps is running, new missing deps may be discovered,
+        // in which case they will keep being added to metadata.discovered
+        for (const id in metadata.discovered) {
+          if (!newData.optimized[id]) {
+            addOptimizedDepInfo(newData, 'discovered', metadata.discovered[id])
+          }
+        }
 
-        // update ssr externals
-        if (ssr) {
-          server._ssrExternals = resolveSSRExternal(
-            server.config,
-            Object.keys(newData.optimized)
-          )
+        // If we don't reload the page, we need to keep browserHash stable
+        if (!needsReload) {
+          newData.browserHash = metadata.browserHash
+          for (const dep in newData.chunks) {
+            newData.chunks[dep].browserHash = metadata.browserHash
+          }
+          for (const dep in newData.optimized) {
+            newData.optimized[dep].browserHash = (
+              metadata.optimized[dep] || metadata.discovered[dep]
+            ).browserHash
+          }
         }
 
-        // While optimizeDeps is running, new missing deps may be discovered,
-        // in which case they will keep being added to metadata.discovered
-        for (const o of Object.keys(metadata.discovered)) {
-          if (!newData.optimized[o]) {
-            newData.discovered[o] = metadata.discovered[o]
+        // Commit hash and needsInterop changes to the discovered deps info
+        // object. Allow for code to await for the discovered processing promise
+        // and use the information in the same object
+        for (const o in newData.optimized) {
+          const discovered = metadata.discovered[o]
+          if (discovered) {
+            const optimized = newData.optimized[o]
+            discovered.browserHash = optimized.browserHash
+            discovered.fileHash = optimized.fileHash
+            discovered.needsInterop = optimized.needsInterop
+            discovered.processing = undefined
           }
         }
-        metadata = server._optimizeDepsMetadata = newData
 
+        metadata = optimizedDeps.metadata = newData
         resolveEnqueuedProcessingPromises()
       }
 
-      if (!processingResult.alteredFiles) {
+      if (!needsReload) {
         commitProcessing()
 
-        logger.info(colors.green(`✨ new dependencies pre-bundled...`), {
+        logger.info(colors.green(`✨ dependencies pre-bundled...`), {
```

Comment:
micro nitpick: Do we need the `...` since we've technically already finished work?

Reply:
I agree. What do you think if we already define better messages?
I'm thinking on:

a. When a page reload isn't needed, it doesn't matter if it is an initial optimization, a missing dependency, or a server restart after a config change, we show:
```
12:18:24 AM [vite] ✨ dependencies pre-bundled: vue, pinia, luxon
```
b. When a page reload is needed:
```
12:18:24 AM [vite] ✨ dependencies pre-bundled: react
12:18:24 AM [vite] ✨ pre-bundled dependencies were outdated, reloading page...
```

I don't think that other messages are useful to the user. It could be a bit confusing why a new `dependencies optimized` line appears after a config change, but we aren't showing that info now.

Is `dependencies optimized` easier to understand than `dependencies pre-bundled`? We use `pre-bundled` in the docs... so probably good to keep that, but with the `pre-bundling` happening after the server, now the `pre` part is a bit strange.

Reply:
Done a thttps://github.com/vitejs/vite/pull/7379/commits/26e726e38272f96bac999bf9f8a143bd46031565

This is how it looks now:
<img width="799" alt="image" src="https://user-images.githubusercontent.com/583075/159453479-302a288f-917a-43f4-b17b-d7492574fd9d.png">

The second line is a missing dep, we could add a `new ` there, but I don't know if it helps, I like that things are aligned. For react-emotion, it would be good to add these to `optimizeDeps.include`

Normally it will look like:
<img width="599" alt="image" src="https://user-images.githubusercontent.com/583075/159455028-e9dc182d-0563-4f8c-8f3e-a8a60f75dbf6.png">